// ssb-uri2/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.decompose = exports.compose = exports.isSSBURI = exports.isExperimentalSSBURIWithAction = exports.isExperimentalSSBURI = exports.isIdentityFusionSSBURI = exports.isIdentityPOBoxSSBURI = exports.isEncryptionKeyBox2DMDiffieHellmanSSBURI = exports.isAddressSSBURI = exports.isBlobSSBURI = exports.isGabbyGroveV1MessageSSBURI = exports.isBendyButtV1MessageSSBURI = exports.isMessageSSBURI = exports.isGabbyGroveV1FeedSSBURI = exports.isBendyButtV1FeedSSBURI = exports.isFeedSSBURI = exports.toMultiserverAddress = exports.toBlobSigil = exports.toMessageSigil = exports.toFeedSigil = exports.fromMultiserverAddress = exports.fromBlobSigil = exports.fromMessageSigil = exports.fromFeedSigil = void 0;
const urlParse = require('url-parse');
const Base64 = {
    unsafeToSafe(input) {
        return input.replace(/\+/g, '-').replace(/\//g, '_');
    },
    safeToUnsafe(input) {
        return input.replace(/-/g, '+').replace(/_/g, '/');
    },
};
function extractBase64Data(pathname) {
    var _a;
    if (!pathname)
        return null;
    const lastPortion = (_a = /(:|\/)([\w_\-=]+)$/.exec(pathname)) === null || _a === void 0 ? void 0 : _a[2];
    if (!lastPortion)
        return null;
    return Base64.safeToUnsafe(lastPortion);
}
function fromFeedSigil(sigil) {
    const data = Base64.unsafeToSafe(sigil.slice(1, -8));
    return `ssb:feed/ed25519/${data}`;
}
exports.fromFeedSigil = fromFeedSigil;
function fromMessageSigil(sigil) {
    const data = Base64.unsafeToSafe(sigil.slice(1, -7));
    return `ssb:message/sha256/${data}`;
}
exports.fromMessageSigil = fromMessageSigil;
function fromBlobSigil(sigil) {
    const data = Base64.unsafeToSafe(sigil.slice(1, -7));
    return `ssb:blob/sha256/${data}`;
}
exports.fromBlobSigil = fromBlobSigil;
function fromMultiserverAddress(msaddr) {
    const encoded = encodeURIComponent(msaddr);
    return `ssb:address/multiserver?multiserverAddress=${encoded}`;
}
exports.fromMultiserverAddress = fromMultiserverAddress;
function toFeedSigil(uri) {
    if (!isFeedSSBURI(uri))
        return null;
    const base64Data = extractBase64Data(urlParse(uri, true).pathname);
    if (!base64Data)
        return null;
    return `@${base64Data}.ed25519`;
}
exports.toFeedSigil = toFeedSigil;
function toMessageSigil(uri) {
    if (!isMessageSSBURI(uri))
        return null;
    const base64Data = extractBase64Data(urlParse(uri, true).pathname);
    if (!base64Data)
        return null;
    return `%${base64Data}.sha256`;
}
exports.toMessageSigil = toMessageSigil;
function toBlobSigil(uri) {
    const base64Data = extractBase64Data(urlParse(uri, true).pathname);
    if (!base64Data)
        return null;
    return `&${base64Data}.sha256`;
}
exports.toBlobSigil = toBlobSigil;
function toMultiserverAddress(uri) {
    return urlParse(uri, true).query.multiserverAddress;
}
exports.toMultiserverAddress = toMultiserverAddress;
function checkTypeFormat(uri, ...args) {
    if (!uri)
        return false;
    const [type, format] = args;
    return ((uri.startsWith(`ssb:${type}:${format}:`) ||
        uri.startsWith(`ssb:${type}/${format}/`) ||
        uri.startsWith(`ssb://${type}/${format}/`)) &&
        !!extractBase64Data(urlParse(uri, true).pathname));
}
function isFeedSSBURI(uri) {
    return checkTypeFormat(uri, 'feed', 'ed25519');
}
exports.isFeedSSBURI = isFeedSSBURI;
function isBendyButtV1FeedSSBURI(uri) {
    return checkTypeFormat(uri, 'feed', 'bendybutt-v1');
}
exports.isBendyButtV1FeedSSBURI = isBendyButtV1FeedSSBURI;
function isGabbyGroveV1FeedSSBURI(uri) {
    return checkTypeFormat(uri, 'feed', 'gabbygrove-v1');
}
exports.isGabbyGroveV1FeedSSBURI = isGabbyGroveV1FeedSSBURI;
function isMessageSSBURI(uri) {
    return checkTypeFormat(uri, 'message', 'sha256');
}
exports.isMessageSSBURI = isMessageSSBURI;
function isBendyButtV1MessageSSBURI(uri) {
    return checkTypeFormat(uri, 'message', 'bendybutt-v1');
}
exports.isBendyButtV1MessageSSBURI = isBendyButtV1MessageSSBURI;
function isGabbyGroveV1MessageSSBURI(uri) {
    return checkTypeFormat(uri, 'message', 'gabbygrove-v1');
}
exports.isGabbyGroveV1MessageSSBURI = isGabbyGroveV1MessageSSBURI;
function isBlobSSBURI(uri) {
    return checkTypeFormat(uri, 'blob', 'sha256');
}
exports.isBlobSSBURI = isBlobSSBURI;
function isAddressSSBURI(uri) {
    var _a;
    if (!uri)
        return false;
    return ((uri.startsWith('ssb:address:multiserver') ||
        uri.startsWith('ssb:address/multiserver') ||
        uri.startsWith('ssb://address/multiserver')) &&
        !!((_a = urlParse(uri, true).query) === null || _a === void 0 ? void 0 : _a.multiserverAddress));
}
exports.isAddressSSBURI = isAddressSSBURI;
function isEncryptionKeyBox2DMDiffieHellmanSSBURI(uri) {
    return checkTypeFormat(uri, 'encryption-key', 'box2-dm-dh');
}
exports.isEncryptionKeyBox2DMDiffieHellmanSSBURI = isEncryptionKeyBox2DMDiffieHellmanSSBURI;
function isIdentityPOBoxSSBURI(uri) {
    return checkTypeFormat(uri, 'identity', 'po-box');
}
exports.isIdentityPOBoxSSBURI = isIdentityPOBoxSSBURI;
function isIdentityFusionSSBURI(uri) {
    return checkTypeFormat(uri, 'identity', 'fusion');
}
exports.isIdentityFusionSSBURI = isIdentityFusionSSBURI;
function isExperimentalSSBURI(uri) {
    if (!uri)
        return false;
    return (uri.startsWith('ssb:experimental') || uri.startsWith('ssb://experimental'));
}
exports.isExperimentalSSBURI = isExperimentalSSBURI;
function isExperimentalSSBURIWithAction(action) {
    return (uri) => {
        var _a;
        return (isExperimentalSSBURI(uri) && ((_a = urlParse(uri, true).query) === null || _a === void 0 ? void 0 : _a.action) === action);
    };
}
exports.isExperimentalSSBURIWithAction = isExperimentalSSBURIWithAction;
function isSSBURI(uri) {
    return (isFeedSSBURI(uri) ||
        isBendyButtV1FeedSSBURI(uri) ||
        isGabbyGroveV1FeedSSBURI(uri) ||
        isMessageSSBURI(uri) ||
        isBendyButtV1MessageSSBURI(uri) ||
        isGabbyGroveV1MessageSSBURI(uri) ||
        isBlobSSBURI(uri) ||
        isAddressSSBURI(uri) ||
        isEncryptionKeyBox2DMDiffieHellmanSSBURI(uri) ||
        isIdentityPOBoxSSBURI(uri) ||
        isIdentityFusionSSBURI(uri) ||
        isExperimentalSSBURI(uri));
}
exports.isSSBURI = isSSBURI;
function validateParts(parts) {
    if (!parts.type)
        throw new Error('Missing required "type" property');
    if (!parts.format)
        throw new Error('Missing required "format" property');
    if (!parts.data)
        throw new Error('Missing required "data" property');
    if (parts.type === 'feed') {
        if (parts.format !== 'ed25519' && parts.format !== 'bendybutt-v1' && parts.format !== 'gabbygrove-v1') {
            throw new Error('Unknown format for type "feed": ' + parts.format);
        }
        else
            return;
    }
    if (parts.type === 'message') {
        if (parts.format !== 'sha256' && parts.format !== 'bendybutt-v1' && parts.format !== 'gabbygrove-v1') {
            throw new Error('Unknown format for type "message": ' + parts.format);
        }
        else
            return;
    }
    if (parts.type === 'blob') {
        if (parts.format !== 'sha256') {
            throw new Error('Unknown format for type "blob": ' + parts.format);
        }
        else
            return;
    }
    if (parts.type === 'address') {
        if (parts.format !== 'multiserver') {
            throw new Error('Unknown format for type "address": ' + parts.format);
        }
        else
            return;
    }
    if (parts.type === 'encryption-key') {
        if (parts.format !== 'box2-dm-dh') {
            throw new Error('Unknown format for type "encryption-key": ' + parts.format);
        }
        else
            return;
    }
    if (parts.type === 'identity') {
        if (parts.format !== 'po-box' && parts.format !== 'fusion') {
            throw new Error('Unknown format for type "identity": ' + parts.format);
        }
        else
            return;
    }
}
function compose(parts) {
    validateParts(parts);
    const { type, format, data } = parts;
    return `ssb:${type}/${format}/${Base64.unsafeToSafe(data)}`;
}
exports.compose = compose;
function decompose(uri) {
    const pathname = urlParse(uri, true).pathname;
    if (!pathname) {
        throw new Error('Invalid SSB URI: ' + uri);
    }
    let [type, format, data] = pathname.split('/');
    data = Base64.safeToUnsafe(data);
    const parts = { type, format, data };
    validateParts(parts);
    return parts;
}
exports.decompose = decompose;
