// ssb-typescript/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function isMsg(msg) {
    return msg && msg.key && msg.value && typeof msg.value === 'object';
}
exports.isMsg = isMsg;
function isRootMsg(msg) {
    var _a, _b, _c;
    return !((_c = (_b = (_a = msg) === null || _a === void 0 ? void 0 : _a.value) === null || _b === void 0 ? void 0 : _b.content) === null || _c === void 0 ? void 0 : _c.root);
}
exports.isRootMsg = isRootMsg;
function isReplyMsgToRoot(rootKey) {
    return function (msg) { var _a, _b, _c; return ((_c = (_b = (_a = msg) === null || _a === void 0 ? void 0 : _a.value) === null || _b === void 0 ? void 0 : _b.content) === null || _c === void 0 ? void 0 : _c.root) === rootKey; };
}
exports.isReplyMsgToRoot = isReplyMsgToRoot;
function isIndirectReplyMsgToRoot(rootKey) {
    return function (msg) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j;
        return ((_c = (_b = (_a = msg) === null || _a === void 0 ? void 0 : _a.value) === null || _b === void 0 ? void 0 : _b.content) === null || _c === void 0 ? void 0 : _c.root) === rootKey ||
            ((_f = (_e = (_d = msg) === null || _d === void 0 ? void 0 : _d.value) === null || _e === void 0 ? void 0 : _e.content) === null || _f === void 0 ? void 0 : _f.branch) === rootKey ||
            ((_j = (_h = (_g = msg) === null || _g === void 0 ? void 0 : _g.value) === null || _h === void 0 ? void 0 : _h.content) === null || _j === void 0 ? void 0 : _j.fork) === rootKey;
    };
}
exports.isIndirectReplyMsgToRoot = isIndirectReplyMsgToRoot;
function isPostMsg(msg) {
    var _a, _b, _c;
    return ((_c = (_b = (_a = msg) === null || _a === void 0 ? void 0 : _a.value) === null || _b === void 0 ? void 0 : _b.content) === null || _c === void 0 ? void 0 : _c.type) === 'post';
}
exports.isPostMsg = isPostMsg;
function isRootPostMsg(msg) {
    var _a, _b, _c;
    return isPostMsg(msg) && !((_c = (_b = (_a = msg) === null || _a === void 0 ? void 0 : _a.value) === null || _b === void 0 ? void 0 : _b.content) === null || _c === void 0 ? void 0 : _c.root);
}
exports.isRootPostMsg = isRootPostMsg;
function isReplyPostMsg(msg) {
    var _a, _b, _c;
    return isPostMsg(msg) && !!((_c = (_b = (_a = msg) === null || _a === void 0 ? void 0 : _a.value) === null || _b === void 0 ? void 0 : _b.content) === null || _c === void 0 ? void 0 : _c.root);
}
exports.isReplyPostMsg = isReplyPostMsg;
function isAboutMsg(msg) {
    var _a, _b, _c;
    return ((_c = (_b = (_a = msg) === null || _a === void 0 ? void 0 : _a.value) === null || _b === void 0 ? void 0 : _b.content) === null || _c === void 0 ? void 0 : _c.type) === 'about';
}
exports.isAboutMsg = isAboutMsg;
function isContactMsg(msg) {
    var _a, _b, _c;
    return ((_c = (_b = (_a = msg) === null || _a === void 0 ? void 0 : _a.value) === null || _b === void 0 ? void 0 : _b.content) === null || _c === void 0 ? void 0 : _c.type) === 'contact';
}
exports.isContactMsg = isContactMsg;
function isVoteMsg(msg) {
    var _a, _b, _c;
    return ((_c = (_b = (_a = msg) === null || _a === void 0 ? void 0 : _a.value) === null || _b === void 0 ? void 0 : _b.content) === null || _c === void 0 ? void 0 : _c.type) === 'vote';
}
exports.isVoteMsg = isVoteMsg;
function isPrivate(msg) {
    var _a, _b, _c, _d;
    if ((_a = msg.meta) === null || _a === void 0 ? void 0 : _a.private)
        return true;
    if (msg.value.private)
        return true;
    if (Array.isArray((_b = msg.value.content) === null || _b === void 0 ? void 0 : _b.recps))
        return true;
    return typeof ((_d = (_c = msg) === null || _c === void 0 ? void 0 : _c.value) === null || _d === void 0 ? void 0 : _d.content) === 'string';
}
exports.isPrivate = isPrivate;
function isPublic(msg) {
    var _a, _b, _c, _d;
    if ((_a = msg.meta) === null || _a === void 0 ? void 0 : _a.private)
        return false;
    if (msg.value.private)
        return false;
    if (Array.isArray((_b = msg.value.content) === null || _b === void 0 ? void 0 : _b.recps))
        return false;
    return typeof ((_d = (_c = msg) === null || _c === void 0 ? void 0 : _c.value) === null || _d === void 0 ? void 0 : _d.content) !== 'string';
}
exports.isPublic = isPublic;
